synapse/storage/databases/main/event_push_actions.py

```diff
@@ -233,14 +233,29 @@ def _get_unread_counts_by_pos_txn(
 
         counts = NotifCounts()
 
-        # First we pull the counts from the summary table
+        # First we pull the counts from the summary table.
+        #
+        # We check that `last_receipt_stream_ordering` matches the stream
+        # ordering given, if it doesn't then a new read receipt has arrived and
+        # we haven't yet updated the counts in `event_push_summary` to reflect
+        # that. In the latter case we simply ignore `event_push_summary` counts
+        # and do a manual count of rows in `event_push_actions` table below.
```

Comment:
Sorry, I still don't get why this is an ok thing to do. Haven't we deleted rows from `event_push_actions`, so relying on it (rather than including `event_push_summary`) will give numbers that are too low?

Reply:
We delete rows at the same time as we update the `event_push_summary`. After a user sends a receipt in the room there are two cases:

1. The receipt has been processed in the `_handle_new_receipts_for_notifs_txn` background task, and so we have deleted the push actions but updated the `event_push_summary`.
2. The receipt has not been processed, so `event_push_summary` is stale *but* we haven't yet deleted the push actions so we can correctly calculate the count from `event_push_actions`.

Maybe:

> and do a manual count of rows in `event_push_actions` table below.
> This gives the correct result as we won't have deleted the push actions yet.

?

Reply:
right, but suppose:
 * a user has 5 unread messages in a room (say E1...E5)
 * A RR arrives for E2. We process it, update `event_push_summary` to say there are 3 remaining unread messages, and delete all the entries from `event_push_actions`.
 * A RR arrives for E4, but before we process it fully, we do this query. We ignore `event_push_summary` and there is nothing left in `event_push_actions`. So we return 0 unread messages instead of 1?

Or am I fundamentally misunderstanding how this works?

Reply:
We delete all entries from `event_push_actions` from *before* E2, so there will still be the three push actions for the three unread messages in there. So in the final step we'd still have push actions E3, E4 and E5 in the `event_push_actions` table